src/inputfetcher.h

```diff
@@ -0,0 +1,197 @@
+        Input(const COutPoint& o LIFETIMEBOUND) noexcept : outpoint{o} {}
+    };
+    std::vector<Input> m_inputs{};
+
+    /**
+     * The set of txids of all txs in the block being fetched.
+     * Used to filter out inputs that are created and spent in the same block,
+     * since they will not be in the db or the cache.
+     */
+    std::unordered_set<Txid, SaltedTxidHasher> m_txids{};
+
+    //! DB coins view to fetch from.
+    const CCoinsView* m_db{nullptr};
+    //! The cache to fetch from.
+    const CCoinsViewCache* m_cache{nullptr};
+
+    std::vector<std::thread> m_worker_threads{};
+    std::barrier<> m_barrier;
+    bool m_request_stop{false};
+
+    void WorkLoop() noexcept
+    {
+        while (true) {
+            m_barrier.arrive_and_wait();
+            if (m_request_stop) [[unlikely]] {
+                return;
+            }
+            while (true) {
+                const size_t i{m_input_head.fetch_add(1, std::memory_order_relaxed)};
+                if (i >= m_inputs.size()) [[unlikely]] {
+                    break;
+                }
+                auto& input{m_inputs[i]};
+                // If an input spends an outpoint from earlier in the block,
+                // it won't be in the cache yet but it also won't be in the db either.
+                if (m_txids.contains(input.outpoint.hash)) {
+                    input.status.store(Input::Status::SKIPPED, std::memory_order_relaxed);
+                    continue;
+                }
+                try {
+                    if (auto coin{m_cache->GetPossiblySpentCoinFromCache(input.outpoint)}) {
+                        input.coin = std::move(*coin);
+                        if (!input.coin.IsSpent()) [[likely]] { // Coin from cache could be spent
+                            // We need release here, so setting coin 2 lines above happens before the main thread loads.
+                            input.status.store(Input::Status::READY, std::memory_order_release);
+                            continue;
+                        }
+                    } else if (auto coin{m_db->GetCoin(input.outpoint)}) {
+                        input.coin = std::move(*coin); // Coin from db cannot be spent, it does not store spent coins
+                        // We need release here, so setting coin in the previous line happens before the main thread loads.
+                        input.status.store(Input::Status::READY, std::memory_order_release);
+                        continue;
+                    }
+                } catch (const std::runtime_error& e) {
+                    LogPrintLevel(BCLog::VALIDATION, BCLog::Level::Warning, "InputFetcher failed to fetch input: %s.\n", e.what());
+                }
+                // Input missing or spent. This block will fail validation.
+                // Skip remaining inputs.
+                m_input_head.store(m_inputs.size(), std::memory_order_relaxed);
```

Comment:
I understand that this poison pill broadcast is needed to stop execution when errors occur - but I'm not sure we should care here about fetching error, I think we should just try to continue if that makes the code simpler

Reply:
In `ConnectBlock` if we encounter a missing input we abort validation immediately. It is wasted work to continue. Why should we do it here?